Fors corrects Elem.apply

som-snytt opened the pull request:
Examples clean-up.  The goal is for each example to compile cleanly under
-unchecked -deprecation -feature
and also run in a meaningful way.

I guess Josh is the implicit reviewer.

Please comment if one example per pull req is what you want, or batched together.

// May the fors be with you...

som-snytt committed:
Fors corrects Elem.apply

jsuereth commented:
Yeap, I'll review now.

jsuereth commented on examples/src/main/scala/fors.scala:
```diff
@@ -2,10 +2,8 @@ package examples
-
-  val e = Node.NoAttributes
+// May the fors be with you...
+object Fors {
```
any reason to prefer capital over lower-case?

jsuereth commented on examples/src/main/scala/fors.scala:
```diff
@@ -39,57 +37,52 @@ object fors {
+  def findAuthorOf(what: String) = searchBooks(books, what, "title", "author")
+  def findTitleBy(who: String) = searchBooks(books, who, "author", "title")
+
+  def removeDuplicates[A](xs: List[A]): List[A] =
```
Why don't use the `distinct` method on `List`?  Wasn't there when this was originally written.

jsuereth commented:
One incredibly minor comment, otherwise LGTM.   If you have a chance to fix up the removeDuplciates, I'll pull.

som-snytt commented:
On API: some examples are superseded by API, though they still demonstrate the concept/usage of fors; another example is computeserver where you'd say just use future with an appropriate executor.  How do you think?

On style: I was about to use caps for object names and type params.  I thought that style evolved over time.  Tell me I'm wrong?

And let me know if you want a big pull or several to many individuals; I can't tell which is easier for you; it looks like there are no competitors to polish these examples, so no conflicts.

One other question: sbt didn't use my scalacOptions (so I'm not using ~ compile).  Can you tell me quickly what incantation?  Not a big deal.

som-snytt commented:
Just remembered, I had a conflict on Windows with upper/lower case.  Don't quote me if it's not a known issue, I didn't investigate fully: with a class BoundedBuffer and object boundedbuffer, I get NoClassDefFound when running.  It looks like the module class file overwrites the BoundedBuffer version.  That's sound too antique to be true.  This is Windows 7, not Windows 95.

jsuereth commented:
On style:   I personally don't like to distinguish between packages, vals and objects.   They're all very similar.   However, companion objects have to share the name of a class.   It's a debatable thing.  I'm ok either way, just curious what your conscious though there was.

On API - If we're showing of _for_ expressions, I think using the collections API is ok, since it doesn't really touch an interesting for.  It's ok to keep the for, but we should also promote the tail-recursion pattern using `@tailrec` too.

As for SBT, you'll have to show me where you set scalacOptions.  build.sbt isn't for the examples project by default.

I prefer small pulls.  This one was perfect and thank you A TON for the help here!  I'm swamped with so much stuff I can't even begin to think about examples, and I can't afford not to as well.

som-snytt committed:
Tweaks to fors, as discussed

Convention for examples is lowerCase object name as container,
which makes the class dir easier to look at, e.g. all classes
for fors are fors$...

jsuereth commented on examples/src/main/scala/fors.scala:
```diff
@@ -39,57 +37,64 @@ object fors {
+    if (xs.isEmpty) xs
+    else xs.head :: removeDuplicatesStackful(for (x <- xs.tail if x != xs.head) yield x)
+
+  // as in List.distinct
```
Love the note!

jsuereth commented on examples/src/main/scala/fors.scala:
```diff
@@ -39,57 +37,64 @@ object fors {
+    import collection.mutable.{HashSet, ListBuffer}
+    val buf = ListBuffer[A]()
+    val seen = HashSet[A]()
+    for (x <- xs)
```
What about:  

```
for(x <- xs; if !seen(x)) {
  buf += x
  seen += x
}
```

?

som-snytt replied:
My fingers almost did that and then stopped after deleting braces from distinct.  I don't mind further suggestions, BTW, I'll apply them.  Examples should absolutely show a cleanest model; otherwise we nulis get the wrong idea and won't unlearn it for years.  (Years in java time, maybe months in Scala time.)

jsuereth commented:
merging as is, we can discuss the style vs. performance for-looping and fix if needed later.
